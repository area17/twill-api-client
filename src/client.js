import { stringify } from 'qs'
import { ofetch } from 'ofetch'

export const client = () => {
  return async (url, fetchOptions = {}) => {
    const headers = {
      ...(fetchOptions.headers || {})
    }

    if (fetchOptions.params) {
      url = `${url}${url.includes('?') ? '&' : '?'}${stringify(
        fetchOptions.params
      )}`
      delete fetchOptions.params
    }

    return await ofetch(url, {
      responseType: 'json',
      retry: 0,
      ...fetchOptions,
      headers
    }).catch((error) => {
      throw error?.data?.errors || error
    })
  }
}
