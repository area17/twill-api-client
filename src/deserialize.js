import { camelCase, isEmpty } from 'lodash-es'
import { camelCaseKeys } from './utils/camel-case-keys'

const MAX_DEPTH = 6

function deserializeRelationships(resources = [], store, depth) {
  return resources
    .map((resource) => deserializeRelationship(resource, store, depth))
    .filter((resource) => !!resource)
}

function deserializeRelationship(resource = {}, store, depth) {
  if (
    typeof resource === 'object' &&
    resource?.type &&
    store[camelCase(resource.type)] &&
    store[camelCase(resource.type)][resource.id]
  ) {
    return deserialize(
      store[camelCase(resource.type)][resource.id],
      store,
      depth
    )
  }

  return null
}

export function deserialize(result, store, depth) {
  if (Array.isArray(result)) {
    return result.map((item) => deserialize(item, store, depth))
  }

  const { id, type, attributes, relationships, meta } =
    store[camelCase(result.type)][result.id]

  if (typeof depth == 'number') {
    depth++
  } else {
    depth = 1
  }

  /**
   * If max depth is reached, simply return the serialized resource
   */
  if (depth > MAX_DEPTH) {
    return { id, type, attributes, relationships, meta }
  }

  let resource = {
    type,
    meta,
    id,
    ...camelCaseKeys(attributes)
  }

  if (relationships) {
    try {
      resource = Object.keys(relationships).reduce((resource, key) => {
        return {
          ...resource,
          [camelCase(key)]: !isEmpty(relationships[key].data)
            ? Array.isArray(relationships[key].data)
              ? deserializeRelationships(relationships[key].data, store, depth)
              : deserializeRelationship(relationships[key].data, store, depth)
            : relationships[key].data
        }
      }, resource)
    } catch (err) {
      console.log(err)
    }
  }

  return resource
}
