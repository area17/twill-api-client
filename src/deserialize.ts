import { camelCase, isEmpty } from 'lodash-es'
import {
  Resource,
  ID,
  JsonApiRelatedResource,
  JsonApiResource,
  NormalizedStore,
} from '@/types'
import { camelCaseKeys } from '@/utils/camel-case-keys'

const MAX_DEPTH = 6

function deserializeRelationships(
  resources: JsonApiRelatedResource[],
  store: NormalizedStore,
  depth: number,
  callback?: (resource: Resource) => void,
): Resource[] {
  return resources
    .map((resource) =>
      deserializeRelationship(resource, store, depth, callback),
    )
    .filter((resource) => !!resource)
}

function deserializeRelationship(
  resource: JsonApiRelatedResource,
  store: NormalizedStore,
  depth: number,
  callback?: (resource: Resource) => void,
): Resource {
  return deserialize(resource, store, depth, callback) as Resource
}

export function deserializeMany<Type extends { type: string; id: ID }>(
  result: Type[],
  store: NormalizedStore,
  depth = 0,
  callback?: (resource: Resource) => void,
): Resource[] {
  return result.map((resource) =>
    deserializeOne(resource, store, depth, callback),
  )
}

export function deserializeOne<Type extends { type: string; id: ID }>(
  result: Type,
  store: NormalizedStore,
  depth = 0,
  callback?: (resource: Resource) => void,
): Resource {
  const serializedResource: JsonApiResource =
    store[camelCase(result.type)][result.id]

  if (typeof depth == 'number') {
    depth++
  } else {
    depth = 1
  }

  /**
   * If max depth is reached, simply return the serialized resource
   */
  if (depth > MAX_DEPTH) {
    return { ...serializedResource } as Resource
  }

  let resource: Resource = {
    type: serializedResource.type,
    meta: serializedResource.meta,
    id: serializedResource.id,
    ...camelCaseKeys(serializedResource.attributes),
  }

  if (serializedResource.relationships) {
    resource = Object.keys(serializedResource.relationships).reduce(
      (acc, key) => {
        const data = serializedResource.relationships[key].data
        let relationship

        if (!isEmpty(data)) {
          relationship = Array.isArray(data)
            ? deserializeRelationships(data, store, depth, callback)
            : deserializeRelationship(data, store, depth, callback)
        }

        return {
          ...acc,
          [camelCase(key)]: relationship ?? data,
        }
      },
      resource,
    )
  }

  if (callback) {
    resource = { ...resource, ...(callback(resource) as unknown as Resource) }
  }

  return resource
}

export function deserialize<Type extends { type: string; id: ID }>(
  result: Type[] | Type,
  store: NormalizedStore,
  depth = 0,
  callback?: (resource: Resource) => void,
): Resource | Resource[] {
  if (Array.isArray(result)) {
    return deserializeMany(result, store, depth, callback)
  }
  return deserializeOne(result, store, depth, callback)
}
