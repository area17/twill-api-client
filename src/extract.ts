import { Resource, Transformations } from '@/types'
import { blocks } from '@/extract/blocks'
import { browsers } from '@/extract/related-items'
// @ts-ignore
import { images } from '@/extract/images'

export const extract = (resource: Resource) => {
  const extracted: Transformations = {}

  if (Array.isArray(resource.blocks)) {
    extracted.blocks = blocks(resource)
  }

  if (Array.isArray(resource.media)) {
    extracted.media = images(resource)
  }

  if (Array.isArray(resource.relatedItems)) {
    extracted.relatedItems = browsers(resource)
  }

  if (Array.isArray(resource.files)) {
    extracted.files = {} // not implemented yet
  }

  return extracted
}
