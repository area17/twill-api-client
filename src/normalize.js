import { camelCase } from 'lodash-es';
import { camelCaseKeys } from '@/utils/camelCaseKeys';

/**
 * Normalize the JSON:API response into categories by resource type
 * and indexed by the resource's ID.
 */
export function normalize(response) {
  let data;

  if (Array.isArray(response.data)) {
    data = response.data;
  } else {
    data = [response.data];
  }

  const included = response.included || [];

  let result = [];
  let resources = {};

  data.forEach((resource) => {
    addResult(result, resource);
    addResource(resources, resource);
  });

  included.forEach((resource) => {
    addResource(resources, resource);
  });

  return {
    result,
    resources
  };
}

function addResult(result, resource) {
  const { type, id } = resource;

  result.push({ type, id });
}

function addResource(resources, resource) {
  const { type, id, attributes, meta, links, relationships } = resource;

  const resourceType = camelCase(type);

  if (!resources[resourceType]) resources[resourceType] = {};

  resources[resourceType][id] = {
    id,
    type,
    attributes: camelCaseKeys(attributes || {}),
    relationships: extractRelationships(relationships),
    meta,
    links
  };

  return resources;
}

function extractRelationships(responseRelationships) {
  if (!responseRelationships) {
    return undefined;
  }

  let relationships = {};

  Object.keys(responseRelationships).map((type) => {
    const relationshipType = camelCase(type);
    relationships[relationshipType] = responseRelationships[type];
  });

  return relationships;
}
