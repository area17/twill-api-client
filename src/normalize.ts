import { camelCase } from 'lodash-es'
import { camelCaseKeys } from './utils/camel-case-keys'
import {
  NormalizedResult,
  NormalizedStore,
  JsonApiRelationship,
  JsonApiDataResponse,
  JsonApiResource,
  JsonApiRelatedResource,
} from './types'

/**
 * Normalize the JSON:API response into categories by resource type
 * and indexed by the resource's ID.
 */
export function normalize(response: JsonApiDataResponse): {
  result: NormalizedResult[]
  resources: NormalizedStore
} {
  let data

  if (Array.isArray(response.data)) {
    data = response.data
  } else {
    data = [response.data]
  }

  const included = response.included || []

  const result: NormalizedResult[] = []
  const resources: NormalizedStore = {}

  data.forEach((resource) => {
    addResult(result, resource)
    addResource(resources, resource)
  })

  included.forEach((resource) => {
    addResource(resources, resource)
  })

  return {
    result,
    resources,
  }
}

function addResult(result: NormalizedResult[], resource: JsonApiResource) {
  const { type, id } = resource

  result.push({ type, id })
}

function addResource(resources: NormalizedStore, resource: JsonApiResource) {
  const { type, id, attributes, meta, links, relationships } = resource

  const resourceType = camelCase(type)

  if (!resources[resourceType]) resources[resourceType] = {}

  resources[resourceType][id] = {
    id,
    type,
    attributes: camelCaseKeys(attributes || {}),
    relationships: extractRelationships(relationships) || {},
    meta,
    links,
  }

  return resources
}

function extractRelationships(responseRelationships: {
  [key: string]: JsonApiRelationship<JsonApiRelatedResource>
}): Record<string, JsonApiRelationship<JsonApiRelatedResource>> | undefined {
  if (!responseRelationships) {
    return undefined
  }

  const relationships: Record<
    string,
    JsonApiRelationship<JsonApiRelatedResource>
  > = {}

  Object.keys(responseRelationships).map((type) => {
    const relationshipType = camelCase(type)
    relationships[relationshipType] = responseRelationships[type]
  })

  return relationships
}
