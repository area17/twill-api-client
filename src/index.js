import { QueryBuilder } from './QueryBuilder.js'
import { normalize } from './normalize.js'
import { deserialize } from './deserialize'
import { allImages } from './extract/images.js'
import { allBlocks } from './extract/blocks.js'

const Twill = ({ url, prefix, version, token }) => {
  const baseURL = `${url}${prefix}/${version}`

  const headers = {
    Authorization: `Bearer ${token}`,
    Accept: 'application/vnd.api+json'
  }

  const get = (path) => {
    return new QueryBuilder({ path, headers })
  }

  const find = (resource) =>
    new QueryBuilder({
      path: resource,
      baseURL,
      headers
    })

  const findOne = (resource, id) =>
    new QueryBuilder({
      path: `${resource}/${id}`,
      baseURL,
      headers
    })

  const findRelated = (resource, response) => {
    const path = response.relationships[resource].links.related

    return new QueryBuilder({ path, headers })
  }

  const findRelationship = (resource, response) => {
    const path = response.relationships[resource].links.self

    return new QueryBuilder({ path, headers })
  }

  const extract = (resource) => {
    return {
      images: allImages(resource),
      editors: allBlocks(resource)
    }
  }

  const transform = (response) => {
    const normalized = normalize(response)
    const resources = deserialize(normalized.result, normalized.resources)
    return resources.map((resource) => {
      return {
        ...resource,
        ...extract(resource)
      }
    })
  }

  return {
    get,
    find,
    findOne,
    findRelated,
    findRelationship,
    normalize,
    deserialize,
    extract,
    transform
  }
}

export { Twill }
