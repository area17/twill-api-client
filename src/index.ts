import { QueryBuilder } from './query-builder'
import {
  ID,
  JsonApiDataResponse,
  JsonApiResource,
  DeserializedResource
} from './types'
import { normalize } from './normalize'
import { deserialize } from './deserialize'
import { blocks } from './extract/blocks'
// @ts-ignore
import { images } from './extract/images'

export interface TwillOptions {
  url: string
  token: string
  prefix: string
  version: string
}

export const Twill = (options: TwillOptions) => {
  const { url, prefix, version, token } = options
  const baseURL = `${url}${prefix}/${version}`

  const headers = {
    Authorization: `Bearer ${token}`,
    Accept: 'application/vnd.api+json'
  }

  const get = (path: string): QueryBuilder => {
    return new QueryBuilder({ path, headers })
  }

  const find = (resource: string) =>
    new QueryBuilder({
      path: resource,
      baseURL,
      headers
    })

  const findOne = (resource: string, id: ID | number | number) =>
    new QueryBuilder({
      path: `${resource}/${id}`,
      baseURL,
      headers
    })

  const findRelated = (
    resource: string,
    response: JsonApiResource
  ): QueryBuilder | null => {
    const path = response.relationships[resource]?.links?.related

    if (!path) {
      return null
    }

    return new QueryBuilder({ path, headers })
  }

  const findRelationship = (
    resource: string,
    response: JsonApiResource
  ): QueryBuilder | null => {
    const path = response.relationships[resource]?.links?.self

    if (!path) {
      return null
    }

    return new QueryBuilder({ path, headers })
  }

  const extract = (resource: DeserializedResource) => {
    return {
      images: images(resource),
      editors: blocks(resource)
    }
  }

  const transform = (response: JsonApiDataResponse) => {
    const normalized = normalize(response)
    const resources: DeserializedResource[] = deserialize(
      normalized.result,
      normalized.resources
    ) as DeserializedResource[]
    return resources.map((resource: DeserializedResource) => {
      return {
        ...resource,
        ...extract(resource)
      }
    })
  }

  return {
    get,
    find,
    findOne,
    findRelated,
    findRelationship,
    normalize,
    deserialize,
    extract,
    transform
  }
}
