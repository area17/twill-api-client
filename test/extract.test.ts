import { expect, test } from 'vitest'
import pages from './data/pages.json'
import { normalize } from '../src/normalize'
import { deserialize } from '../src/deserialize'
import { extract } from '../src/extract'

test('extract files', () => {
  const normalized = normalize(pages)
  const deserialized = deserialize(normalized.result, normalized.resources)
  const extracted = extract(deserialized[0])
  const deserializedWithCallback = deserialize(
    normalized.result,
    normalized.resources,
    0,
    extract,
  )

  expect(Object.keys(extracted)).toMatchInlineSnapshot(`
  [
    "blocks",
    "files",
  ]
  `)

  expect(extracted).toMatchSnapshot()

  expect(deserializedWithCallback).toMatchSnapshot()
})
