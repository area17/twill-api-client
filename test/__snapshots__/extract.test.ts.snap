// Vitest Snapshot v1

exports[`extract 2`] = `
{
  "blocks": {
    "default": [
      {
        "blockType": "interactive",
        "blocks": undefined,
        "childKey": null,
        "content": {
          "colorTheme": "none",
        },
        "editorName": "default",
        "files": [],
        "id": "1",
        "media": [],
        "meta": {
          "browsers": {
            "interactive": [
              1,
            ],
          },
        },
        "position": 1,
        "relatedItems": [
          {
            "browserName": "interactive",
            "id": "1",
            "meta": undefined,
            "position": 1,
            "related": {
              "blocks": undefined,
              "createdAt": "2022-08-08T22:22:24.000000Z",
              "displayTitle": "Interactive module",
              "files": undefined,
              "id": "1",
              "media": undefined,
              "meta": undefined,
              "published": true,
              "title": "Interactive module",
              "type": "interactives",
              "updatedAt": "2022-08-08T22:25:12.000000Z",
            },
            "type": "related-items",
          },
        ],
        "type": "blocks",
      },
      {
        "blockType": "text",
        "blocks": undefined,
        "childKey": null,
        "content": {
          "colorTheme": "dark-gray",
          "text": "<p>Incididunt qui duis elit ut et proident esse ullamco id dolor. Id esse fugiat est consectetur excepteur proident ex ullamco anim. Incididunt aliquip consequat labore ipsum esse. Officia irure consequat quis non labore eiusmod esse ea mollit eiusmod occaecat do culpa.</p>",
        },
        "editorName": "default",
        "files": [],
        "id": "2",
        "media": [],
        "meta": {
          "browsers": null,
        },
        "position": 2,
        "relatedItems": [],
        "type": "blocks",
      },
      {
        "blockType": "images",
        "blocks": undefined,
        "childKey": null,
        "content": {
          "body": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla quam morbi lectus nunc, nam tellus sed porttitor nunc. Urna neque purus eget id ut eu aenean feugiat diam.",
          "colorTheme": "dark-gray",
          "display": "image-right",
          "title": "Images",
        },
        "editorName": "default",
        "files": [],
        "id": "3",
        "media": [
          {
            "alt": "test-image",
            "caption": "",
            "createdAt": "2022-11-21T15:32:08.000000Z",
            "crop": "default",
            "filename": "test-image.png",
            "height": 1486,
            "id": "1",
            "lqip": null,
            "meta": {
              "crop": "default",
              "role": "images",
              "uuid": "f2bdab3c-2e04-4318-b0eb-7d86b71aa1ec/test-image.png",
            },
            "metadata": [],
            "originalSrc": "https://twill.test/storage/uploads/f2bdab3c-2e04-4318-b0eb-7d86b71aa1ec/test-image.png",
            "ratio": "default",
            "role": "images",
            "src": "https://twill.test/img/f2bdab3c-2e04-4318-b0eb-7d86b71aa1ec/test-image.png?fm=jpg&q=80&fit=max&crop=1485%2C1486%2C46%2C0",
            "type": "media",
            "updatedAt": "2022-11-21T15:32:08.000000Z",
            "uuid": "f2bdab3c-2e04-4318-b0eb-7d86b71aa1ec/test-image.png",
            "video": "",
            "width": 1485,
          },
        ],
        "meta": {
          "browsers": null,
        },
        "position": 3,
        "relatedItems": [],
        "type": "blocks",
      },
    ],
  },
  "files": {},
}
`;

exports[`extract 3`] = `
[
  {
    "blocks": {
      "default": [
        {
          "blockType": "interactive",
          "blocks": undefined,
          "childKey": null,
          "content": {
            "colorTheme": "none",
          },
          "editorName": "default",
          "files": {},
          "id": "1",
          "media": {},
          "meta": {
            "browsers": {
              "interactive": [
                1,
              ],
            },
          },
          "position": 1,
          "relatedItems": {
            "interactive": [
              {
                "blocks": undefined,
                "createdAt": "2022-08-08T22:22:24.000000Z",
                "displayTitle": "Interactive module",
                "files": undefined,
                "id": "1",
                "media": undefined,
                "meta": undefined,
                "published": true,
                "title": "Interactive module",
                "type": "interactives",
                "updatedAt": "2022-08-08T22:25:12.000000Z",
              },
            ],
          },
          "type": "blocks",
        },
        {
          "blockType": "text",
          "blocks": undefined,
          "childKey": null,
          "content": {
            "colorTheme": "dark-gray",
            "text": "<p>Incididunt qui duis elit ut et proident esse ullamco id dolor. Id esse fugiat est consectetur excepteur proident ex ullamco anim. Incididunt aliquip consequat labore ipsum esse. Officia irure consequat quis non labore eiusmod esse ea mollit eiusmod occaecat do culpa.</p>",
          },
          "editorName": "default",
          "files": {},
          "id": "2",
          "media": {},
          "meta": {
            "browsers": null,
          },
          "position": 2,
          "relatedItems": {},
          "type": "blocks",
        },
        {
          "blockType": "images",
          "blocks": undefined,
          "childKey": null,
          "content": {
            "body": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla quam morbi lectus nunc, nam tellus sed porttitor nunc. Urna neque purus eget id ut eu aenean feugiat diam.",
            "colorTheme": "dark-gray",
            "display": "image-right",
            "title": "Images",
          },
          "editorName": "default",
          "files": {},
          "id": "3",
          "media": {
            "images": [
              {
                "default": {
                  "alt": "test-image",
                  "caption": "",
                  "createdAt": "2022-11-21T15:32:08.000000Z",
                  "crop": "default",
                  "filename": "test-image.png",
                  "height": 1486,
                  "id": "1",
                  "lqip": null,
                  "meta": {
                    "crop": "default",
                    "role": "images",
                    "uuid": "f2bdab3c-2e04-4318-b0eb-7d86b71aa1ec/test-image.png",
                  },
                  "metadata": [],
                  "originalSrc": "https://twill.test/storage/uploads/f2bdab3c-2e04-4318-b0eb-7d86b71aa1ec/test-image.png",
                  "ratio": "default",
                  "role": "images",
                  "src": "https://twill.test/img/f2bdab3c-2e04-4318-b0eb-7d86b71aa1ec/test-image.png?fm=jpg&q=80&fit=max&crop=1485%2C1486%2C46%2C0",
                  "type": "media",
                  "updatedAt": "2022-11-21T15:32:08.000000Z",
                  "uuid": "f2bdab3c-2e04-4318-b0eb-7d86b71aa1ec/test-image.png",
                  "video": "",
                  "width": 1485,
                },
              },
            ],
          },
          "meta": {
            "browsers": null,
          },
          "position": 3,
          "relatedItems": {},
          "type": "blocks",
        },
      ],
    },
    "colorTheme": "light-gray",
    "description": "Ipsum sunt culpa occaecat Lorem aute proident eu aliquip id. Qui aute consequat et sit irure velit magna. Cupidatat tempor commodo qui aute adipisicing adipisicing tempor commodo occaecat laborum irure deserunt. Ullamco quis enim et amet ex fugiat exercitation.",
    "files": {},
    "hero": {
      "assetType": "video",
      "blendMode": "default",
    },
    "id": "1",
    "media": undefined,
    "meta": undefined,
    "pies": undefined,
    "published": true,
    "slug": "page-slug",
    "title": "Lorem ipsum dolor sit amet",
    "type": "pages",
  },
]
`;
